cmd/bee-join/main.go

```diff
@@ -0,0 +1,158 @@
+		Path:   "bzz-chunk",
+	}
+	return &apiStore{
+		baseUrl: u.String(),
+	}, nil
+}
+
+// Get implements storage.Getter
+func (a *apiStore) Get(ctx context.Context, mode storage.ModeGet, address swarm.Address) (ch swarm.Chunk, err error) {
+	addressHex := address.String()
+	url := strings.Join([]string{a.baseUrl, addressHex}, "/")
+	res, err := http.DefaultClient.Get(url)
+	if err != nil {
+		return nil, err
+	}
+	if res.StatusCode != http.StatusOK {
+		return nil, fmt.Errorf("chunk %s not found", addressHex)
+	}
+	chunkData, err := ioutil.ReadAll(res.Body)
+	if err != nil {
+		return nil, err
+	}
+	ch = swarm.NewChunk(address, chunkData)
+	return ch, nil
+}
+
+// Join is the underlying procedure for the CLI command
+func Join(cmd *cobra.Command, args []string) (err error) {
+	// if output file is specified, create it if it does not exist
+	var outFile *os.File
+	if outFilePath != "" {
+		// make sure we have full path
+		outDir := filepath.Dir(outFilePath)
+		if outDir != "." {
+			err := os.MkdirAll(outDir, 0o777) // skipcq: GSC-G301
+			if err != nil {
+				return err
+			}
+		}
+		// protect any existing file unless explicitly told not to
+		outFileFlags := os.O_CREATE | os.O_WRONLY
+		if outFileForce {
+			outFileFlags |= os.O_TRUNC
+		} else {
+			outFileFlags |= os.O_EXCL
+		}
+		// open the file
+		outFile, err = os.OpenFile(outFilePath, outFileFlags, 0o666) // skipcq: GSC-G302
+		if err != nil {
+			return err
+		}
+		defer outFile.Close()
+	} else {
+		outFile = os.Stdout
+	}
+
+	// process the reference to retrieve
+	addr, err := swarm.ParseHexAddress(args[0])
+	if err != nil {
+		return err
+	}
+
+	// initialize interface with HTTP API
+	store, err := newApiStore(host, port, ssl)
+	if err != nil {
+		return err
+	}
+
+	// create the join and get its data reader
+	j := joiner.NewSimpleJoiner(store)
+	r, l, err := j.Join(context.Background(), addr)
+	if err != nil {
+		return err
+	}
+
+	// join, rinse, repeat until done
+	teeReader := io.TeeReader(r, outFile)
+	data := make([]byte, swarm.ChunkSize)
+	var total int64
+	for i := int64(0); i < l; i += swarm.ChunkSize {
+		c, err := teeReader.Read(data)
+		if err != nil {
+			return err
+		}
+		total += int64(c)
+	}
+	if total != l {
+		return fmt.Errorf("received only %d of %d total bytes", total, l)
+	}
+	return nil
+}
+
+func main() {
+	c := &cobra.Command{
+		Use:   "join [hash]",
+		Args:  cobra.ExactArgs(1),
+		Short: "Retrieve data from Swarm",
+		Long: `Assembles chunked data from referenced by a root Swarm Hash.
+
+Will output retrieved data to stdout.`,
+		RunE: Join,
+	}
+	c.Flags().StringVarP(&outFilePath, "output-file", "o", "", "file to write output to")
+	c.Flags().BoolVarP(&outFileForce, "force", "f", false, "overwrite existing output file")
+	c.Flags().StringVar(&host, "host", "127.0.0.1", "api host")
+	c.Flags().IntVar(&port, "port", 8080, "api port")
+	c.Flags().BoolVar(&ssl, "ssl", false, "use ssl")
+
+	err := c.Execute()
+	if err != nil {
+		os.Exit(1)
```

Comment:
maybe print the error before exiting?

Reply:
It gets printed already when I run it with error. I think `cobra` takes care of it internally.

Reply:
I find too verbose for cobra to print usage when the error happens, so I would suggest to set `SilenceUsage: true` on the Command, if you agree, not required. Errors are still printed, just without the usage, while `-h` works correctly.